components/eamxx/src/diagnostics/tests/horiz_avg_test.cpp

```diff
@@ -0,0 +1,197 @@
+  diag1->initialize(t0, RunType::Initial);
+
+  diag1->compute_diagnostic();
+  auto diag1_f = diag1->get_diagnostic();
+
+  FieldIdentifier diag0_fid("qc_horiz_avg_manual",
+                            scalar1d_layout.clone().strip_dim(COL), kg / kg,
+                            grid->name());
+  Field diag0(diag0_fid);
+  diag0.allocate_view();
+  auto diag0_v = diag0.get_view<Real>();
+
+  auto qc1_v  = qc1.get_view<Real *>();
+  auto area_v = area.get_view<const Real *>();
+
+  // calculate total area
+  Real atot = field_sum<Real>(area, &comm);
+  // calculate weighted avg
+  Real wavg = sp(0.0);
+  Kokkos::parallel_reduce(
+      "HorizAvgDiag::compute_diagnostic_impl::weighted_sum", ngcols,
+      KOKKOS_LAMBDA(const int icol, Real &local_wavg) {
+        local_wavg += (area_v[icol] / atot) * qc1_v[icol];
+      },
+      wavg);
+  Kokkos::deep_copy(diag0_v, wavg);
+
+  diag1_f.sync_to_host();
+  auto diag1_v_h = diag1_f.get_view<Real, Host>();
+  REQUIRE(diag1_v_h() == wavg);
+
+  // Try known cases
+  // Set qc1_v to 1.0 to get weighted average of 1.0
+  wavg = sp(1.0);
+  Kokkos::deep_copy(qc1_v, wavg);
+  diag1->compute_diagnostic();
+  auto diag1_v2_host = diag1_f.get_view<Real, Host>();
+  REQUIRE_THAT(diag1_v2_host(),
+               Catch::Matchers::WithinRel(
+                   wavg, tol));  // Catch2's floating point comparison
+
+  // other diags
+  // Set qc2_v to 5.0 to get weighted average of 5.0
+  wavg       = sp(5.0);
+  auto qc2_v = qc2.get_view<Real **>();
+  Kokkos::deep_copy(qc2_v, wavg);
+
+  diag2->set_required_field(qc2);
+  diag2->initialize(t0, RunType::Initial);
+  diag2->compute_diagnostic();
+  auto diag2_f = diag2->get_diagnostic();
+
+  auto diag2_v_host = diag2_f.get_view<Real *, Host>();
+
+  for(int i = 0; i < nlevs; ++i) {
+    REQUIRE_THAT(diag2_v_host(i), Catch::Matchers::WithinRel(wavg, tol));
+  }
+
+  auto qc3_v = qc3.get_view<Real ***>();
+  FieldIdentifier diag3_manual_fid("qc_horiz_avg_manual",
+                                   scalar3d_layout.clone().strip_dim(COL),
+                                   kg / kg, grid->name());
+  Field diag3_manual(diag3_manual_fid);
+  diag3_manual.allocate_view();
+  auto diag3_manual_v = diag3_manual.get_view<Real **>();
+  // calculate diag3_manual by hand
+  auto p = ESU::get_default_team_policy(dim3 * nlevs, ngcols);
+  Kokkos::parallel_for(
+      "HorizAvgDiag::compute_diagnostic_impl::manual_diag3", p,
+      KOKKOS_LAMBDA(const TeamMember &m) {
+        const int idx = m.league_rank();
+        const int j   = idx / nlevs;
+        const int k   = idx % nlevs;
+        Real sum      = sp(0.0);
+        Kokkos::parallel_reduce(
+            Kokkos::TeamThreadRange(m, ngcols),
+            [&](const int icol, Real &accum) {
+              accum += (area_v(icol) / atot) * qc3_v(icol, j, k);
+            },
+            sum);
+        Kokkos::single(Kokkos::PerTeam(m),
+                       [&]() { diag3_manual_v(j, k) = sum; });
+      });
```

Comment:
I have one possible suggestion, feel free to discard.

What about using horiz_contraction in the test as well? It may make for a much easier/shorter test to maintain...